samples/HttpServer_FirmwareUpload/app/application.cpp

```diff
@@ -37,44 +147,35 @@ void onFile(HttpRequest& request, HttpResponse& response)
 
 int onUpload(HttpServerConnection& connection, HttpRequest& request, HttpResponse& response)
 {
+	Serial.printf("onUpload\n");
 	ReadWriteStream* file = request.files["firmware"];
 	if(file == nullptr) {
 		debug_e("Something went wrong with the file upload");
 		return 1;
 	}
 
-	debugf("Uploaded length: %d", file->available());
-
-	rboot_config bootConfig = rboot_get_config();
-	uint8_t slot = bootConfig.current_rom;
-	slot = (slot == 0 ? 1 : 0);
-	Serial.printf("Firmware updated, rebooting to rom %d...\r\n", slot);
-	rboot_set_current_rom(slot);
-	System.restart(5); // defer the restart with 5 seconds to give time to the web server to return the response
-
-	response.sendFile("restart.html");
-
+	SignedRbootOutputStream* uploadStream = static_cast<SignedRbootOutputStream*>(file);
+	uploadStream->close();
```

Comment:
Don't close the stream here. It will be closed later at the end of the request.

Reply:
`uploadStream->close()` triggers the signature validation and therefore must be invoked *before* returning from `onUpdate` to decided if the received image is valid and the slot can be activated.
Did I miss something? According to https://github.com/SmingHub/Sming/blob/develop/docs/source/experimental/signed-ota.rst, the `close` method is the intended place for signature verification.

Reply:
> uploadStream->close() triggers the signature validation

Not necessarily . It would be better if there is a `validate()` method that triggers the validation and sets the ok flag. Also the ok() method can be renamed to isValid(). The closing of the file is something separate and does not need to be coupled with the validation.  You can move the logic that deletes the first sector if the validation is not correct to the destructor.

> the close method is the intended place for signature verification.

I will change the documentation later in the week.

Reply:
Moved validation code out of the `close` method. Still don't like it, because the `onUpload` callback should inform the user that *all* of the firmware update process has been completed successfully. Even if the signature matches, writing the flash can still fail, and as the last chunk of flash is only written in `RbootOutputStream::close` that must come before preparing the user response. 
I also don't like than an upcast from `ReadWriteStream` to `SignedRbootOutputStream` is necessary in `onUpload`. It may be an acceptable compromise for this example, but for a potential future component, I would prefer a more type-safe solution.

Reply:
> I also don't like than an upcast from ReadWriteStream to SignedRbootOutputStream is necessary in onUpload. It may be an acceptable compromise for this example, but for a potential future component, I would prefer a more type-safe solution.

Feel free to suggest better solution as PR. This API is not final.

Reply:
I'll try to come up with something, but it may take a couple of weeks.

What's your opinion on the current state of the example. Is it acceptable for this PR?

Reply:
> What's your opinion on the current state of the example. Is it acceptable for this PR?

You are doing great and I am happy to welcome you to our team and thank you for your first accepted contribution :)